Source/Tools/LinuxEmulation/LinuxSyscalls/SeccompEmulator.cpp

```diff
@@ -0,0 +1,890 @@
+    const auto Offset = Inst->k;
+    if (Offset & 0b11) {
+      ///< Wasn't 4-byte aligned.
+      return -EINVAL;
+    }
+
+    if (Offset >= sizeof(seccomp_data)) {
+      ///< Tried accessing outside of seccomp_data.
+      return -EINVAL;
+    }
+
+    ldr(DestReg, REG_SECCOMP_DATA, Offset);
+    break;
+  }
+  case BPF_MEM:
+    if (Inst->k >= 16) {
+      ///< Larger than scratch space size.
+      return -EINVAL;
+    }
+
+    ldr(DestReg, REG_SECCOMP_DATA, offsetof(WorkingBuffer, ScratchMemory[Inst->k]));
+    break;
+  case BPF_IND:
+  case BPF_MSH: return -EINVAL; ///< Unsupported
+  case BPF_LEN:
+    ///< Just returns the length of seccomp_data.
+    movz(DestReg, sizeof(seccomp_data));
+    break;
+  }
+
+  return 0;
+}
+
+uint64_t BPFEmitter::HandleStore(uint32_t BPFIP, const sock_filter* Inst) {
+  switch (BPF_SIZE(Inst->code)) {
+  case BPF_H:
+  case BPF_B:
+  case 0x18: /* BPF_DW */ return -EINVAL;
+  case BPF_W: break;
+  }
+
+  const auto SrcReg = BPF_CLASS(Inst->code) == BPF_LD ? REG_A : REG_X;
+  if (Inst->k >= 16) {
+    ///< Larger than scratch space size.
+    return -EINVAL;
+  }
+
+  str(SrcReg, REG_SECCOMP_DATA, offsetof(WorkingBuffer, ScratchMemory[Inst->k]));
+
+  return 0;
+}
+
+uint64_t BPFEmitter::HandleALU(uint32_t BPFIP, const sock_filter* Inst) {
+  const auto SrcType = BPF_SRC(Inst->code);
+  const auto Op = BPF_OP(Inst->code);
+
+  switch (Op) {
+  case BPF_ADD:
+  case BPF_SUB:
+  case BPF_MUL:
+  case BPF_DIV:
+  case BPF_OR:
+  case BPF_AND:
+  case BPF_LSH:
+  case BPF_RSH:
+  case BPF_MOD:
+  case BPF_XOR: {
+    auto SrcReg = REG_X;
+    if (SrcType == BPF_K) {
+      SrcReg = REG_TMP;
+      movz(SrcReg, Inst->k & 0xFFFF);
+      movk(SrcReg, Inst->k >> 16, 16);
+    }
+
+    switch (Op) {
+    case BPF_ADD: add(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg); break;
+    case BPF_SUB: sub(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg); break;
+    case BPF_MUL: mul(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg); break;
+    case BPF_DIV:
+      ///< Specifically unsigned.
+      udiv(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg);
+      break;
+    case BPF_OR: orr(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg); break;
+    case BPF_AND: and_(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg); break;
+    case BPF_LSH: lslv(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg); break;
+    case BPF_RSH: lsrv(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg); break;
+    case BPF_MOD:
+      ///< Specifically unsigned.
+      udiv(ARMEmitter::Size::i32Bit, REG_TMP2, REG_A, SrcReg);
+      msub(ARMEmitter::Size::i32Bit, REG_A, REG_TMP2, SrcReg, REG_A);
+      break;
+    case BPF_XOR: eor(ARMEmitter::Size::i32Bit, REG_A, REG_A, SrcReg); break;
+    default: return -EINVAL;
+    }
+
+    break;
+  }
+  case BPF_NEG:
+    if (SrcType == BPF_X) {
+      ///< Only BPF_K supported on NEG.
+      return -EINVAL;
+    }
+
+    mvn(ARMEmitter::Size::i32Bit, REG_A, REG_A);
```

Comment:
I believe this should be `neg`, not `mvn`. (Arithmetic inverse instead of bitwise inverse)

Reply:
Per the reference implementation - https://github.com/torvalds/linux/blob/master/arch/arm64/net/bpf_jit_comp.c#L970-L974

Reply:
You're right. I'll need to fix that.